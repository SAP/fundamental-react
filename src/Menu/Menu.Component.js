import React from 'react';
import { Description, DocsText, DocsTile, Header, Import, Properties, Separator } from '../_playground';
import { Menu, MenuGroup, MenuItem, MenuList } from '../';

export const MenuComponent = () => {
    const menuCode = `<Menu>
    <MenuList>
        <MenuItem url="/">Option 1</MenuItem>
        <MenuItem url="/">Option 2</MenuItem>
        <MenuItem url="/">Option 3</MenuItem>
        <MenuItem url="/">Option 4</MenuItem>
    </MenuList>
</Menu>`;

    const menuGroupCode = `<Menu>
    <MenuList>
        <MenuItem link="/">Option 1</MenuItem>
        <MenuItem link="/">Option 2</MenuItem>
        <MenuItem link="/">Option 3</MenuItem>
    </MenuList>
    <MenuGroup title="Group Header">
        <MenuList>
            <MenuItem link="/">Option 4</MenuItem>
            <MenuItem link="/">Option 5</MenuItem>
            <MenuItem link="/">Option 6</MenuItem>
        </MenuList>
    </MenuGroup>
</Menu>`;

    const menuSeparatorCode = `<Menu>
    <MenuList>
        <MenuItem link="/" separator>
            Option 1
        </MenuItem>
        <MenuItem link="/" separator>
            Option 2
        </MenuItem>
        <MenuItem link="/" separator>
            Option 3
        </MenuItem>
        <MenuItem link="/">Option 4</MenuItem>
    </MenuList>
</Menu>`;

    const menuAddonBeforeCode = `<Menu addonBefore={true}>
    <MenuList>
        <MenuItem link="/">Option 1</MenuItem>
        <MenuItem link="/" addon="accept">Option 2</MenuItem>
        <MenuItem link="/">Option 3</MenuItem>
        <MenuItem link="/">Option 4</MenuItem>
    </MenuList>
</Menu>`;

    return (
        <div>
            <Header>Menu</Header>
            <Description>
                The menu component is the listing structure with optional headers to create menus. Commonly used as the
                contents when composing “dropdowns”, “contextual menus”, “mega menu”, etc, when paired with the popover
                component.
            </Description>
            <Import module='Menu, MenuList, MenuItem, MenuGroup' path='/fundamental-react/src/' />

            <Separator />

            <Properties
                properties={[
                    { name: 'url', description: 'string - href attribute of <a> tag. Use either \'url\' or \'link\'' },
                    { name: 'link', description: 'string - a router link. Use either \'url\' or \'link\'' },
                    { name: 'separator', description: 'bool - when set to true, adds a horizontal line (separator).' },
                    {
                        name: 'title',
                        description: 'string - group header title.'
                    },
                    {
                        name: 'addonBefore',
                        description: 'bool - when set to true, enables menu items with addon before.'
                    },
                    {
                        name: 'addon',
                        description: 'string - the name of the SAP icon to be applied as an addon before.'
                    }
                ]}
                type='Inputs' />

            <Separator />

            <h2>Menu</h2>
            <p>The basic stucture of a menu.</p>
            <DocsTile>
                <Menu>
                    <MenuList>
                        <MenuItem url='/'>Option 1</MenuItem>
                        <MenuItem url='/'>Option 2</MenuItem>
                        <MenuItem url='/'>Option 3</MenuItem>
                        <MenuItem url='/'>Option 4</MenuItem>
                    </MenuList>
                </Menu>
            </DocsTile>
            <DocsText>{menuCode}</DocsText>
            <Separator />

            <h2>Menu w/ Group</h2>
            <p>Menu with grouped sub-menus and group headers.</p>
            <DocsTile>
                <Menu>
                    <MenuList>
                        <MenuItem link='/'>Option 1</MenuItem>
                        <MenuItem link='/'>Option 2</MenuItem>
                        <MenuItem link='/'>Option 3</MenuItem>
                    </MenuList>
                    <MenuGroup title='Group Header'>
                        <MenuList>
                            <MenuItem link='/'>Option 4</MenuItem>
                            <MenuItem link='/'>Option 5</MenuItem>
                            <MenuItem link='/'>Option 6</MenuItem>
                        </MenuList>
                    </MenuGroup>
                </Menu>
            </DocsTile>
            <DocsText>{menuGroupCode}</DocsText>
            <Separator />

            <h2>Menu w/ Separator</h2>
            <p>Menu items with horizontal line as separator.</p>
            <DocsTile>
                <Menu>
                    <MenuList>
                        <MenuItem link='/' separator>
                            Option 1
                        </MenuItem>
                        <MenuItem link='/' separator>
                            Option 2
                        </MenuItem>
                        <MenuItem link='/' separator>
                            Option 3
                        </MenuItem>
                        <MenuItem link='/'>Option 4</MenuItem>
                    </MenuList>
                </Menu>
            </DocsTile>
            <DocsText>{menuSeparatorCode}</DocsText>
            <Separator />

            <h2>Menu w/ Addon Before</h2>
            <p>Menu items with an addon before.</p>
            <DocsTile>
                <Menu addonBefore>
                    <MenuList>
                        <MenuItem link='/'>Option 1</MenuItem>
                        <MenuItem addon='accept' link='/'>
                            Option 2
                        </MenuItem>
                        <MenuItem link='/'>Option 3</MenuItem>
                        <MenuItem link='/'>Option 4</MenuItem>
                    </MenuList>
                </Menu>
            </DocsTile>
            <DocsText>{menuAddonBeforeCode}</DocsText>
            <Separator />
        </div>
    );
};
