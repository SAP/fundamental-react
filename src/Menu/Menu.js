import classnamesBind from 'classnames/bind';
import MenuItem from './_MenuItem';
import MenuList from './_MenuList';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/menu.css';

const classnames = classnamesBind.bind(styles);

/** A **Menu** is the listing structure with optional headers to create menus. Commonly used as the
contents when composing “dropdowns”, “contextual menus”,  etc, when paired with the **Popover**
component. */
const Menu = React.forwardRef(({ children, className, cssNamespace, ...props }, ref) => {

    const menuClasses = classnames(
        `${cssNamespace}-menu`,
        className
    );

    return (
        <nav {...props} className={menuClasses}
            ref={ref}>
            {children}
        </nav>
    );
});

Menu.displayName = 'Menu';

Menu.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string
};

Menu.Item = MenuItem;
Menu.List = MenuList;

export default withStyles(Menu);
