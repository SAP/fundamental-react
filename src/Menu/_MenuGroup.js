import classnames from 'classnames';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import PropTypes from 'prop-types';
import React from 'react';

const MenuGroup = ({ title, children, className, headingLevel, titleProps, ...props }) => {
    const menuGroupClasses = classnames(
        'fd-menu__group',
        className
    );

    const HeadingTag = `h${headingLevel}`;

    return (
        <div {...props} className={menuGroupClasses}>
            <HeadingTag {...titleProps} className='fd-menu__title'>{title}</HeadingTag>
            {children}
        </div>
    );
};

MenuGroup.displayName = 'Menu.Group';

MenuGroup.propTypes = {
    title: PropTypes.string.isRequired,
    children: PropTypes.node,
    className: PropTypes.string,
    headingLevel: CustomPropTypes.range(2, 6),
    titleProps: PropTypes.object
};

MenuGroup.defaultProps = {
    headingLevel: 3
};

export default MenuGroup;
