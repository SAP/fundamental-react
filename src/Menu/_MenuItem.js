import classnamesBind from 'classnames/bind';
import Icon from '../Icon/Icon';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import iconStyles from 'fundamental-styles/dist/icon.css';
import menuItemStyles from 'fundamental-styles/dist/menu.css';

const classnames = classnamesBind.bind({
    ...iconStyles,
    ...menuItemStyles
});

const MenuItem = ({
    addonAfter,
    addonBefore,
    url,
    isLink,
    children,
    onClick,
    className,
    cssNamespace,
    addonProps,
    urlProps,
    separator,
    active,
    disabled,
    selected,
    ...props
}) => {

    const linkClassNames = classnames(
        `${cssNamespace}-menu__link`,
        {
            'is-active': active,
            'is-selected': selected,
            'is-disabled': disabled
        }
    );

    const addonBeforeClassname = classnames(`${cssNamespace}-menu__addon-before`);
    const addonAfterClassname = classnames(`${cssNamespace}-menu__addon-after`);

    const renderLink = () => {
        if (url) {
            return (<a {...urlProps}
                className={linkClassNames}
                href={url}
                onClick={onClick}
                role='menuitem'>
                {addonBefore &&
                    <span className={addonBeforeClassname}>
                        <Icon {...addonProps} ariaHidden
                            glyph={addonBefore} />
                    </span>
                }
                <span className={classnames(`${cssNamespace}-menu__title`)}>{children}</span>
                {addonAfter &&
                    <span className={addonAfterClassname}>
                        <Icon {...addonProps} ariaHidden
                            glyph={addonAfter} />
                    </span>
                }
            </a>);
        } else if (children && React.isValidElement(children)) {
            const childrenClassnames = classnames(
                linkClassNames,
                children.props.className
            );

            return (
                <>
                    {addonBefore &&
                        <span {...urlProps} className={addonBeforeClassname}>
                            <Icon {...addonProps} ariaHidden
                                glyph={addonBefore} />
                        </span>
                    }
                    <span className={classnames(`${cssNamespace}-menu__title`)}>
                        {React.Children.map(children, child => {
                            return React.cloneElement(child, {
                                className: childrenClassnames,
                                ...urlProps
                            });
                        })}
                    </span>
                    {addonAfter &&
                        <span {...urlProps} className={addonAfterClassname}>
                            <Icon {...addonProps} ariaHidden
                                glyph={addonAfter} />
                        </span>
                    }
                </>
            );
        } else if (children) {
            return (<a {...urlProps}
                className={linkClassNames}
                onClick={onClick}
                role='menuitem'>
                {addonBefore &&
                    <span className={addonBeforeClassname}>
                        <Icon {...addonProps} ariaHidden
                            glyph={addonBefore} />
                    </span>
                }
                <span className={classnames(`${cssNamespace}-menu__title`)}>{children}</span>
                {addonAfter &&
                    <span className={addonAfterClassname}>
                        <Icon {...addonProps} ariaHidden
                            glyph={addonAfter} />
                    </span>
                }
            </a>);
        }
    };

    const listClassNames = classnames(
        `${cssNamespace}-menu__item`,
        className
    );

    return (
        <>
            <li
                {...props}
                className={listClassNames}
                role='presentation'>
                {renderLink()}
            </li>
            {separator && <span className={classnames(`${cssNamespace}-menu__separator`)} />}
        </>
    );
};

MenuItem.displayName = 'Menu.Item';

MenuItem.propTypes = {
    /** Set to **true** to apply active style */
    active: PropTypes.bool,
    /** Name of the SAP icon to be applied as an add-on before the text */
    addonAfter: PropTypes.string,
    /** Name of the SAP icon to be applied as an add-on after the text */
    addonBefore: PropTypes.string,
    /** Additional props to be spread to the add-ons */
    addonProps: PropTypes.object,
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Set to **true** to style as a link */
    isLink: PropTypes.bool,
    /** Set to **true** to apply selected style */
    selected: PropTypes.bool,
    /** Set to **true** to place a separator after list item */
    separator: PropTypes.bool,
    /** Enables use of `<a>` element. Value to be applied to the anchor\'s `href` attribute.
     * Should use either `link` or `url`, but not both. */
    url: PropTypes.string,
    /** Additional props to be spread to the Menu Item links (when using `url`). */
    urlProps: PropTypes.object,
    /**
     * Callback function; triggered when the MenuItem (i.e. the `<a>` element) is clicked.
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent. See https://reactjs.org/docs/events.html.
     * @returns {void}
    */
    onClick: PropTypes.func
};

export default withStyles(MenuItem);
