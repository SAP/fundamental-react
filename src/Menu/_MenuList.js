import classnamesBind from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import withStyles from '../utils/withStyles';
import styles from 'fundamental-styles/dist/menu.css';

const classnames = classnamesBind.bind(styles);

const MenuList = ({ children, className, cssNamespace, ...props }) => {
    const menuListClasses = classnames(
        `${cssNamespace}-menu__list`,
        className
    );

    return (
        <ul
            {...props}
            className={menuListClasses}
            role='menu'>
            {children}
        </ul>
    );
};

MenuList.displayName = 'Menu.List';

MenuList.propTypes = {
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string
};

export default withStyles(MenuList);
