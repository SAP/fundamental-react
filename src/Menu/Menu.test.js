import Menu from './Menu';
import { mount } from 'enzyme';
import React from 'react';
import { Link, MemoryRouter } from 'react-router-dom';

describe('<Menu />', () => {
    describe('Prop spreading', () => {
        test('should allow props to be spread to the Menu component', () => {
            const element = mount(<Menu data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MenuList component', () => {
            const element = mount(<Menu.List data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MenuItem component', () => {
            const element = mount(<Menu.Item data-sample='Sample' />);

            expect(
                element.getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MenuItem component\'s addon div element', () => {
            const element = mount(
                <MemoryRouter>
                    <Menu>
                        <Menu.List>
                            <Menu.Item>
                                <Link to='/'>Option 1</Link>
                            </Menu.Item>
                            <Menu.Item addonBefore='accept' addonProps={{ 'data-sample': 'Sample' }}>
                                <Link to='/'>Option 2</Link>
                            </Menu.Item>
                            <Menu.Item>
                                <Link to='/'>Option 3</Link>
                            </Menu.Item>
                            <Menu.Item>
                                <Link to='/'>Option 4</Link>
                            </Menu.Item>
                        </Menu.List>
                    </Menu>
                </MemoryRouter>);

            expect(
                element.find('.sap-icon--accept').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MenuItem component\'s Link component', () => {
            const element = mount(
                <MemoryRouter>
                    <Menu>
                        <Menu.List>
                            <Menu.Item urlProps={{ 'data-sample': 'Sample' }}>
                                <Link to='/' />
                            </Menu.Item>
                        </Menu.List>
                    </Menu>
                </MemoryRouter>);

            expect(
                element.find('a.fd-menu__link').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MenuItem component\'s a element', () => {
            const element = mount(
                <Menu>
                    <Menu.List>
                        <Menu.Item url='/' urlProps={{ 'data-sample': 'Sample' }} />
                    </Menu.List>
                </Menu>
            );

            expect(
                element.find('a.fd-menu__link').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <Menu ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('NAV');
        expect(ref.current.className).toEqual('fd-menu');
    });
});
