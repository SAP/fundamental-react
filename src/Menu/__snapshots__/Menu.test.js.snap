// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> create basic menu component 1`] = `
<nav
  className="fd-menu"
>
  <ul
    className="fd-menu__list"
  >
    <li>
      <a
        className="fd-menu__item"
        href="/"
      >
        Option 1
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item fd-menu__link"
        href="/"
      >
        Option 2
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item"
        href="/"
      >
        Option 3
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item"
        href="/"
      >
        Option 4
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item"
      >
        Option 5
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Menu /> create menu add-on component 1`] = `
<nav
  className="fd-menu fd-menu--addon-before"
>
  <ul
    className="fd-menu__list"
  >
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 1
      </a>
    </li>
    <li>
      <div
        className="fd-menu__addon-before"
      >
        <span
          className="sap-icon--accept"
        />
      </div>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 2
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 3
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 4
      </a>
    </li>
  </ul>
</nav>
`;

exports[`<Menu /> create menu group component 1`] = `
<nav
  className="fd-menu blue"
>
  <ul
    className="fd-menu__list blue"
  >
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 1
      </a>
    </li>
    <li
      className="blue"
    >
      <a
        className="fd-menu__item fd-menu__link"
        href="/"
        onClick={[Function]}
      >
        Option 2
      </a>
    </li>
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 3
      </a>
    </li>
  </ul>
  <div
    className="fd-menu__group blue"
  >
    <h1
      className="fd-menu__title"
    >
      Group Header
    </h1>
    <ul
      className="fd-menu__list"
    >
      <li>
        <a
          className="fd-menu__item"
          href="/"
          onClick={[Function]}
        >
          Option 4
        </a>
      </li>
      <li>
        <a
          className="fd-menu__item"
          href="/"
          onClick={[Function]}
        >
          Option 5
        </a>
      </li>
      <li>
        <a
          className="fd-menu__item"
          href="/"
          onClick={[Function]}
        >
          Option 6
        </a>
      </li>
    </ul>
  </div>
  <div
    className="fd-menu__group"
  >
    <h1
      className="fd-menu__title"
    >
      Group Header 2
    </h1>
    <ul
      className="fd-menu__list"
    >
      <li>
        <a
          className="fd-menu__item"
          href="/"
          onClick={[Function]}
        >
          Option 7
        </a>
      </li>
      <li>
        <a
          className="fd-menu__item"
          href="/"
          onClick={[Function]}
        >
          Option 8
        </a>
      </li>
      <li>
        <a
          className="fd-menu__item"
          href="/"
          onClick={[Function]}
        >
          Option 9
        </a>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`<Menu /> create menu with separator component 1`] = `
<nav
  className="fd-menu"
>
  <ul
    className="fd-menu__list"
  >
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 1
      </a>
    </li>
    <hr />
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 2
      </a>
    </li>
    <hr />
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 3
      </a>
    </li>
    <hr />
    <li>
      <a
        className="fd-menu__item"
        href="/"
        onClick={[Function]}
      >
        Option 4
      </a>
    </li>
  </ul>
</nav>
`;
